Block restrict nodal patch recovery object

recuero opened the pull request:
Closes  #25732.

recuero committed:
Block restrict nodal patch recovery object (#25732)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1820193/) on cd5010b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25733/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1820211/) on cd5010b wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>6760c9</th>
      <th colspan=4>#25733 cd5010</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.73%</td>
      <td>85.74%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>92096</td>
      <td>92100</td>
      <td>+4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15328</td>
      <td>15324</td>
      <td>-4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25733/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25733/coverage/framework)


## Modules coverage

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>6760c9</th>
      <th colspan=4>#25733 cd5010</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.63%</td>
      <td>84.62%</td>
      <td>-0.01%</td>
      <td>71.43%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>25763</td>
      <td>25769</td>
      <td>+6</td>
      <td>10</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>4679</td>
      <td>4683</td>
      <td>+4</td>
      <td>4</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25733/coverage_diff/tensor_mechanics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25733/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25733/coverage/xfem)

</details>

## Warnings

- `tensor_mechanics` new line coverage rate 71.43% is less than the suggested 90.0%



This comment will be updated on new commits.

dschwen commented on modules/tensor_mechanics/src/userobjects/NodalPatchRecoveryBase.C:
```diff
@@ -62,8 +62,8 @@ NodalPatchRecoveryBase::nodalPatchRecovery(const Point & x,
   {
-    A += _Ae.at(elem_id);
-    b += _be.at(elem_id);
+    A += libmesh_map_find(_Ae, elem_id);
```
What's the advantage of libmesh map find over at?

recuero replied:
It gives you a backtrace if it fails. It's also been requested in PR reviews.

hugary1995 replied:
Interesting... Weren't we catching this sort of exception and emitting backtrace before?

recuero replied:
The simulation did stop but I didn't get any backtrace, just that the key was invalid.

recuero committed:
Add 'stateful' mortar constraint test (#25732)

In order to make a stateful constraint we:

 - Use nodal patch recovery to retrieve a nodal value (existing)
 - Use an auxiliary variable to store the recovered quantity (existing)
 - Pull an old state of the nodal variable in a mortar constraint

recuero commented:
@dschwen in the second commit there is an example of a "stateful material property" mortar constraint obtained through nodal patch recovery and then nodal variable evaluation at the mortar quadrature points. That seems to work well, we can use it as a reference.

hugary1995 commented on modules/tensor_mechanics/src/auxkernels/NodalPatchRecoveryAux.C:
```diff
@@ -39,7 +39,20 @@ NodalPatchRecoveryAux::computeValue()
+        if (block_id == _mesh.elemPtr(elem_id)->subdomain_id())
+        {
+          elem_ids.push_back(elem_id);
+          continue;
```
```suggestion
```

hugary1995 commented on modules/tensor_mechanics/src/auxkernels/NodalPatchRecoveryAux.C:
I'd like to see a separate method that takes a vector of elements and returns a vector of block-restricted elements.

hugary1995 commented on modules/combined/test/include/constraints/GapConductanceStatefulConstraint.h:
```diff
@@ -0,0 +1,42 @@
+  std::vector<const ADVariableValue *> _disp_primary;
+  ///@}
+
+  const VariableValue & _stress_old;
```
Could you add docstring for this member?

hugary1995 commented on modules/combined/test/include/constraints/GapConductanceStatefulConstraint.h:
```diff
@@ -0,0 +1,42 @@
+  ///@}
+
+  const VariableValue & _stress_old;
+  const VariableValue & _stress_neighbor_old;
```
Same here.

hugary1995 reviewed (changes requested):
Minor comments.

We might as well think about how to hide the details from the users, i.e. right now we need a userobject plus an auxkernel for each material property, and they should have matching block restrictions (maybe it could be relaxed with some sort of block restriction validation?).

If possible, I think it may be a good idea to let the auxkernel create the corresponding userobject automatically. Instead of letting the user do it.

recuero committed:
Address Gary's review

recuero commented:
@hugary1995 Thanks for the review.

> If possible, I think it may be a good idea to let the auxkernel create the corresponding userobject automatically. Instead of letting the user do it.

I agree. I think we can create an action in the future that would take care of this. We could streamline it so that we can do NPR to many properties using a vector, just like we do in the TM action.

hugary1995 reviewed (changes requested):
A minor comment to consider.

hugary1995 commented on modules/tensor_mechanics/src/userobjects/NodalPatchRecoveryBase.C:
```diff
@@ -174,3 +174,17 @@ NodalPatchRecoveryBase::finalize()
       _communicator, query_ids, gather_data, act_on_data, ex);
 }
+
+bool
+NodalPatchRecoveryBase::isBlockRestrictionSame(
+    const std::set<SubdomainID> & aux_kernel_subdomains) const
+{
+  std::set<SubdomainID> result;
+  std::set_symmetric_difference(blockIDs().begin(),
+                                blockIDs().end(),
+                                aux_kernel_subdomains.begin(),
+                                aux_kernel_subdomains.end(),
+                                std::inserter(result, result.end()));
+
+  return result.size() == 0 ? true : false;
+}
```
oops I missed this. Why don't we just use `operator==`?

recuero replied:
Ended up deleting this method since there is block subset functionality in the parent objects.

recuero commented on modules/tensor_mechanics/include/userobjects/NodalPatchRecoveryBase.h:
```diff
@@ -29,6 +29,16 @@ class NodalPatchRecoveryBase : public ElementUserObject
    */
   virtual Real nodalPatchRecovery(const Point & p, const std::vector<dof_id_type> & elem_ids) const;
 
+  /**
+   * Compares whether a calling object has the same block restrictio as this user object. This is
+   * handy to make sure there is consistency between nodal patch recovery methods in the auxiliary
+   * kernels and the user objects.
+   *
+   * @param aux_kernel_subdomains Set of subdomains on which the external object acts upon.
+   * @return Whether the extenal object and this object act on the same subdomains.
+   */
+  bool isBlockRestrictionSame(const std::set<SubdomainID> & aux_kernel_subdomains) const;
```
> Is this too restrictive? For example, what happens if the userobject is restricted A, B, C, while the auxkernel is restricted to just A, should we error in that case?

Yeah, we can consider this way too restrictive. I was thinking that both aux kernel and user object should match since the purpose is to recover the properties in all the selected blocks. But will change that.

recuero commented:
Related to #24030.

hugary1995 commented on modules/tensor_mechanics/src/auxkernels/NodalPatchRecoveryAux.C:
```diff
@@ -27,6 +27,11 @@ NodalPatchRecoveryAux::validParams()
 {
+  // Check user object block restriction for consistency
+  if (!isBlockSubset(_npr.blockIDs()))
+    paramError("nodal_patch_recovery_uo",
+               "Nodal patch recovery auxiliary kernel is not defined in a subset of blocks of the "
+               "associated auxiliary kernel. Revise your input file.");
```
```suggestion
    paramError("nodal_patch_recovery_uo",
               "Nodal patch recovery auxiliary kernel is not defined in a subset of blocks of the "
               "associated userobject. Revise your input file.");
```

recuero replied:
:(
